[WIP] Add datasets to config api (pipelines)

leoromanovich opened the pull request:
For now, we don't push you to follow any predefined schema of issue, but ensure you've already read our contribution guide: https://open-metric-learning.readthedocs.io/en/latest/from_readme/contributing.html.

leoromanovich commented:
@AlekseySh what do you think about implementation like that?

AlekseySh commented on oml/registry/datasets.py:
```diff
@@ -0,0 +1,43 @@
+from oml.registry.transforms import get_transforms_by_cfg
+
+
+def _build_image_dataset(**cfg: TCfg) -> Tuple[ILabeledDataset, IQueryGalleryLabeledDataset]:
```
let's make naming similar to other modules
`get_image_datasets_by_cfg`
and
`get_image_datasets`

leoromanovich replied:
Done

AlekseySh commented on tests/test_oml/test_registry/test_registry.py:
```diff
@@ -61,6 +66,7 @@ def get_opt() -> Optimizer:
         ("transforms", TRANSFORMS_REGISTRY, get_transforms, None),
         ("pairwise_model", PAIRWISE_MODELS_REGISTRY, get_pairwise_model, None),
         ("postprocessor", POSTPROCESSORS_REGISTRY, get_postprocessor, None),
+        ("dataset", DATASET_BUILDER_REGISTRY, build_image_dataset, {"dataset_root": MOCK_DATASET_PATH}),
```
it should be "datasets_getter" for which you create a folder with example configs
look at the other tests nearby 

I don't understand why this test not failed without this folder of configs. Please, add the folder and update the test so it fails in such cases.

leoromanovich replied:
Oh, i haven't noticed, that datasets folder in gitignore. Yea, I need to add it to git.

AlekseySh replied:
oh
yep, and after that and solving conflicts, you can start making tests green

leoromanovich replied:
Done

AlekseySh commented on oml/lightning/pipelines/logging.py:
```diff
@@ -117,7 +117,9 @@ def log_pipeline_info(self, cfg: TCfg) -> None:
 
         # log dataframe
-        self.run["dataset"].upload(str(Path(cfg["dataset_root"]) / cfg["dataframe_name"]))
+        self.run["dataset"].upload(
```
the same about other dumpable stuff later
(that's what I tried to explain you before when said that ideally we need an Interface to handle such logic)

leoromanovich replied:
I'm not sure if I've done this properly. Actually, I don't like the way i've chosen for now. After all, I believe that special class for dumping staff is a better idea.

leoromanovich commented on tests/test_runs/test_pipelines/configs/train.yaml:
```diff
@@ -4,22 +4,25 @@ accelerator: cpu
 
-dataset_root: path_to_replace  # we will replace it in runtime with the default dataset folder
-dataframe_name: df.csv
+dataset:
```
Not sure about `datasets_getter` naming. We use transform_*, extractor, sampler etc. `datasets` looks more intuitive, what do you think? @AlekseySh

AlekseySh replied:
It highlights that the registry is for dataset getters, not for the dataset. It also highlights that we get two datasets from the same function, so, we can conclude the we can perform some additional steps and we will have access to the both dataset. But you are right, that it looks different from the other methods.

We need one more opinion. @DaloroAT what do u think?

DaloroAT replied:
I will write my suggestion here.

First, the structure of `yaml` for the dataset is the same as Dataset signature:

```yaml
name: oml_image_dataset
args:
  df: df.csv
  dataset_root: null
  transform:
    name: norm_resize_torch
    args:
      im_size: 224
  cache_size: 100
```

And part of full config:

```yaml
datasets:
  train:
    name: image_labeled_dataset  # or `image_query_gallery_labeled_dataset`
    args:
      df: df.csv
      dataset_root: null
      transform:
        name: norm_resize_torch
        args:
          im_size: 224
      cache_size: 0
  validation:
    name: image_query_gallery_labeled_dataset
    args:
      df: df.csv
      dataset_root: null
      transform:
        name: norm_resize_torch
        args:
          im_size: 322
      cache_size: 10000
  predict:
    name: image_base_dataset
    args:
      df: df_predict.csv
      dataset_root: /.../pictures/
      transform:
        name: norm_resize_torch
        args:
          im_size: 512
      cache_size: 0
```

And parser as pseudo:

```python
def get_dataset(name: str, type_dataset: str, **kwargs: Dict[str, Any]):
    constructor = DATASETS_REGISTRY[name]
    df_path = kwargs["df"]

    df = pd.read_csv(...)

    if type_dataset in ["train", "validation"]:
        df = df[df[SPLIT_COLUMN] == type_dataset].reset_index(drop=True)
        
    return constructor(df=df, ...)


def get_datasets_by_cfg(**cfg: TCfg):
    datasets_cfg = cfg["datasets"]
    datasets = {}
    for type_dataset in datasets_cfg.keys():
        datasets[type_dataset] = get_dataset(
            datasets_cfg["name"],
            type_dataset,
            **datasets_cfg["args"],
        )
    return datasets

```

For training extractor and training postprocessing we can check that user specify correct set of keys (at least `train`, or `train` and `validation`).

DaloroAT replied:
I don't like to use `type_dataset` as indicator for split, so we can change it a bit and specify explicitly in yaml, that csv should be splitted (anyway we do it right now implicitly)

```yaml
datasets:
  train:
    name: image_labeled_dataset
    split: train
    args: ...
  validation:
    name: image_query_gallery_labeled_dataset
    split: validation
    args: ...
  predict:
    name: image_base_dataset
    args: ...
```

AlekseySh replied:
answered offline

leoromanovich commented on tests/test_oml/test_registry/test_registry.py:
```diff
@@ -49,6 +50,10 @@ def get_opt() -> Optimizer:
 
 
+def get_mock_dataset_path() -> str:
+    return str(MOCK_DATASET_PATH / "df.csv")
```
that is outdated func. I'll drop that)

leoromanovich replied:
Done

leoromanovich commented on oml/registry/datasets.py:
```diff
@@ -0,0 +1,43 @@
+
+
+DATASET_BUILDER_REGISTRY = {
+    "oml_image_dataset": _build_image_dataset,
```
Changed to `datasets`.
Done

leoromanovich commented on oml/registry/datasets.py:
```diff
@@ -0,0 +1,43 @@
+    return dataset_train, dataset_val
+
+
+def build_image_dataset(
```
Done)

leoromanovich commented on tests/test_runs/test_pipelines/configs/validate.yaml:
```diff
@@ -1,13 +1,16 @@
 
-transforms_val:
-  name: norm_resize_torch
+dataset:
```
For now I've converted everything. I suggest to choose, what config will be remained in old format.

AlekseySh replied:
just randomly pick whatever you want
let it be training with bboxes, for example

leoromanovich commented on oml/lightning/pipelines/parser.py:
```diff
@@ -104,7 +104,30 @@ def parse_ckpt_callback_from_config(cfg: TCfg) -> ModelCheckpoint:
+        print("Used old cfg version. Converting to new format")
+        args = dict()
+        for key in keys_to_process:
+            args[key] = cfg[key]
```
Good one, changed

AlekseySh commented on oml/lightning/pipelines/predict.py:
```diff
@@ -30,12 +30,12 @@ def extractor_prediction_pipeline(cfg: TCfg) -> None:
 
-    transforms = get_transforms_by_cfg(cfg["transforms_predict"])
-    filenames = [list(Path(cfg["data_dir"]).glob(f"**/*.{ext}")) for ext in IMAGE_EXTENSIONS]
+    transforms = get_transforms_by_cfg(cfg["datasets"]["args"]["transforms_predict"])
```
something weird happened here, I think you should not touch this pipeline at the moment

AlekseySh commented:
based on offline discussion: 


```
------------------------
PREDICT.YAML

precision: 32
accelerator: gpu
devices: 1

dataset:
  name: BaseImgDataset
  im_paths: ...  # or im_dir
  transforms_predict:
    name: norm_resize_albu
    args:
      im_size: 224

save_dir: "."

bs: 64
num_workers: 10

extractor:
  name: vit
  args:
    arch: vits16
    normalise_features: False
    use_multi_scale: False
    weights: vits16_cars

hydra:
  run:
    dir: ${save_dir}
  searchpath:
   - pkg://oml.configs
  job:
    chdir: True

-------------------
VALIDATE.YAML

accelerator: gpu
devices: 1
precision: 32


bs_val: 256
num_workers: 8

val_dataset:
  name: image_dataset
  dataframe_name: df_with_bboxes.csv  # df/path_to_df
  args:
    dataset_root: data/CARS196/
    transforms_val:
      name: norm_resize_albu
      args:
        im_size: 224

extractor:
  name: vit
  args:
    arch: vits16
    normalise_features: False
    use_multi_scale: False
    weights: vits16_cars

metric_args:
  metrics_to_exclude_from_visualization: [cmc,]
  cmc_top_k: [1, 5]
  map_top_k: [5]
  precision_top_k: [5]
  fmr_vals: [0.01]
  pcf_variance: [0.5, 0.9, 0.99]
  return_only_overall_category: False
  visualize_only_overall_category: True

hydra:
  searchpath:
   - pkg://oml.configs
  job:
    chdir: True

-----------------------------
REGISTRY.PY

REGISTRY_DATASETS = {
  "image_dataset": ImageQGLDataset
}

def get_dataset_by_cfg(cfg,  split_val=None):
  if split_val and dataframe_name in cfg:
    df = pd.read_csv(cfg["dataframe_name"])
    df = df[df.split == split_val]

  return REGISTRY_DATASETS["image_dataset"](df_path/df)


-------------------------------
PIPELINES.PY


train_dataset = get_dataset_by_cfg(cfg["train_dataset"], split_val="train")
val_dataset = get_dataset_by_cfg(cfg["valid_dataset"], split_val="validate")

```

leoromanovich commented:
Looks like we can't avoid a small builder, because of transforms initialisation and mapping.
from this
```python
REGISTRY_DATASETS = {
  "image_dataset": ImageQGLDataset
}

def get_dataset_by_cfg(cfg,  split_val=None):
  if split_val and dataframe_name in cfg:
    df = pd.read_csv(cfg["dataframe_name"])
    df = df[df.split == split_val]

  return REGISTRY_DATASETS["image_dataset"](df_path/df)
```

to something like that:

```python
REGISTRY_DATASETS = {
  "image_qg_dataset": qg_builder,
}

def qg_builder(args):
    transforms = ...
    dataset = QGDataset(....)
    return dataset
    

def get_dataset_by_cfg(cfg,  split_val=None):
  if split_val and dataframe_name in cfg:
    df = pd.read_csv(cfg["dataframe_name"])
    mapper = {...}
    df = df[df.split == split_val]
    df = df[].map(mapper)
    cfg['df'] = df  # because not all datasets will have a dataframe, we pack it inside cfg.
  return REGISTRY_DATASETS["image_qg_dataset"](df_path/df)
```
